network/transport/grpc/connection.go

```diff
@@ -169,62 +147,105 @@ func (mc *conn) verifyOrSetPeerID(id transport.PeerID) bool {
 }
 
-func (mc *conn) registerClientStream(clientStream grpc.ClientStream, method string) error {
+func (mc *conn) Send(protocol Protocol, envelope interface{}) error {
 	mc.mux.Lock()
 	defer mc.mux.Unlock()
 
-	if _, exists := mc.grpcOutboundStreams[method]; exists {
-		return fmt.Errorf("peer is already connected (method=%s)", method)
+	outbox := mc.outboxes[protocol.MethodName()]
+	if outbox == nil {
+		return fmt.Errorf("can't send message, protocol not connected: %s", protocol.MethodName())
 	}
-
-	// when the gRPC stream cancels, cancel the connection's context
-	if mc.ctx == nil {
-		mc.ctx, mc.cancelCtx = context.WithCancel(context.Background())
+	if len(outbox) >= cap(outbox) {
+		// This node is a slow responder, we'll have to drop this message because our backlog is full.
+		return fmt.Errorf("peer's outbound message backlog has reached max capacity, message is dropped (peer=%s,backlog-size=%d)", mc.Peer(), cap(outbox))
 	}
-	go func(cancel func()) {
-		<-clientStream.Context().Done()
-		cancel()
-	}(mc.cancelCtx)
+	outbox <- envelope
 
-	mc.grpcOutboundStreams[method] = clientStream
 	return nil
 }
 
-func (mc *conn) registerServerStream(serverStream grpc.ServerStream) error {
+func (mc *conn) registerStream(protocol Protocol, stream Stream) bool {
 	mc.mux.Lock()
 	defer mc.mux.Unlock()
 
-	method := grpc.ServerTransportStreamFromContext(serverStream.Context()).Method()
-	if _, exists := mc.grpcInboundStreams[method]; exists {
-		return fmt.Errorf("peer is already connected (method=%s)", method)
-	}
-	if _, exists := mc.grpcOutboundStreams[method]; exists {
-		return fmt.Errorf("peer is already connected (method=%s)", method)
+	methodName := protocol.MethodName()
+	if mc.streams[methodName] != nil {
+		return false
 	}
 
-	mc.grpcInboundStreams[method] = serverStream
-
-	// when the gRPC stream cancels, cancel the connection's context
 	if mc.ctx == nil {
 		mc.ctx, mc.cancelCtx = context.WithCancel(context.Background())
 	}
+
+	mc.streams[methodName] = stream
+	mc.outboxes[methodName] = make(chan interface{}, 20)
+
+	mc.startReceiving(protocol, stream)
+	mc.startSending(protocol, stream)
+
+	// A connection can have multiple active streams, but if one of them is closed, all of them should be closed, also closing the underlying connection.
 	go func(cancel func()) {
-		// Wait for the serverStream to close. Then remove it, and if it was the last one, callback
-		<-serverStream.Context().Done()
+		<-stream.Context().Done()
 		cancel()
-		mc.mux.Lock()
-		defer mc.mux.Unlock()
-		// Remove this stream from the inbound stream registry
-		delete(mc.grpcInboundStreams, method)
-		// If empty, remove.
-		if len(mc.grpcInboundStreams) == 0 {
-			mc.inboundStreamsClosedCallback(mc)
-		}
 	}(mc.cancelCtx)
-	return nil
+
+	return true
 }
 
-func (mc *conn) open(tlsConfig *tls.Config, connectedCallback func(grpcConn *grpc.ClientConn) bool) {
+func (mc *conn) startReceiving(protocol Protocol, stream Stream) {
+	go func() {
```

Comment:
Nitpick: wouldn't it be cleaner to remove the goroutine here and call `go startReceiving(...)` on line `180`?